File: lsl/reader/tbx.cpp
#include "Python.h"
#include <cmath>
#include <cstdint>

#define NO_IMPORT_ARRAY
#define PY_ARRAY_UNIQUE_SYMBOL gofast_ARRAY_API
#include "numpy/arrayobject.h"

#include "readers.hpp"

/*
  TBX Reader
*/

typedef struct __attribute__((packed)) {
    uint32_t syncWord;
    union {
        struct {
            uint32_t frame_count:24;
            uint8_t  frame_id:8;
        };
        uint32_t frame_count_word;
    };
    uint32_t second_count;
    uint32_t first_chan;
    uint32_t nstand;
    uint32_t nchan;
} TBXHeader;


typedef struct __attribute__((packed)) {
    int64_t timetag;
    uint8_t bytes[8192];       // The maximum possible size
} TBXPayload;


typedef struct __attribute__((packed)) {
    TBXHeader header;
    TBXPayload payload;
} TBXFrame;

/*
  State variables to keep track of the last read_tbx_impl used
*/

static int cached_nstand = 256;
static int cached_nchan = 16;

PyObject *tbx_method = NULL;
PyObject *tbx_size = NULL;

template<int NSTAND, int NCHAN, typename T, NPY_TYPES N>
PyObject *read_tbx_impl(PyObject *self, PyObject *args) {
    PyObject *ph, *buffer, *output, *frame, *fHeader, *fPayload, *temp;
    PyArrayObject *data;
    int i, nstand, nchan;
    TBXFrame cFrame;
    
    static constexpr int frameSize = sizeof(TBXHeader)+8+NCHAN*NSTAND*2*1;
    
    if(!PyArg_ParseTuple(args, "OO", &ph, &frame)) {
        PyErr_Format(PyExc_RuntimeError, "Invalid parameters");
        return NULL;
    }
    
    // Read from the file - header + timestamp + NSTAND-sized buffer
    if( tbx_method == NULL ) {
        tbx_method = Py_BuildValue("s", "read");
    }
    if( tbx_size == NULL ) {
        tbx_size = Py_BuildValue("i", frameSize);
    }
    buffer = PyObject_CallMethodObjArgs(ph, tbx_method, tbx_size, NULL);
    if( buffer == NULL ) {
        if( PyObject_HasAttrString(ph, "read") ) {
            PyErr_Format(PyExc_IOError, "An error occured while reading from the file");
        } else {
            PyErr_Format(PyExc_AttributeError, "Object does not have a read() method");
        }
        return NULL;
    } else if( PyBytes_GET_SIZE(buffer) != frameSize ) {
        PyErr_Format(EOFError, "End of file encountered during filehandle read");
        Py_XDECREF(buffer);
        return NULL;
    }
    memcpy(&cFrame, PyBytes_AS_STRING(buffer), frameSize);
    Py_XDECREF(buffer);
    
    // Determine the number of stands and channels in the frame 
    nstand = __bswap_32(cFrame.header.nstand);
    nchan = __bswap_32(cFrame.header.nstand);
    
    // If nstand is not what we expect, update the cache and retry with correct size
    if( nstand != NSTAND || nchan != NCHAN ) {
        cached_nstand = nstand; // Update for next time
        cached_nchan = nchan;
        Py_XDECREF(tbx_size);   // Update for next time
        tbx_size = NULL;        // Force NULL since Py_XDECREF isn't guaranteed to change tbx_size
        PyObject_CallMethod(ph, "seek", "ii", -frameSize, 1);
        
        switch(nstand) {
            case 64:
                switch(nchan) {
                    case 4:
                        return read_tbx_impl<64, 4, T, N>(self, args);
                    case 8:
                        return read_tbx_impl<64, 8, T, N>(self, args);
                    case 12:
                        return read_tbx_impl<64, 12, T, N>(self, args);
                    case 16:
                        return read_tbx_impl<64, 16, T, N>(self, args);
                    default:
                        PyErr_Format(SyncError, "Mark 5C sync word differs from expected");
                        return NULL;
                }
            case 128:
                switch(nchan) {
                    case 4:
                        return read_tbx_impl<128, 4, T, N>(self, args);
                    case 8:
                        return read_tbx_impl<128, 8, T, N>(self, args);
                    case 12:
                        return read_tbx_impl<128, 12, T, N>(self, args);
                    case 16:
                        return read_tbx_impl<128, 16, T, N>(self, args);
                    default:
                        PyErr_Format(SyncError, "Mark 5C sync word differs from expected");
                        return NULL;
                }
            case 256:
                switch(nchan) {
                    case 4:
                        return read_tbx_impl<256, 4, T, N>(self, args);
                    case 8:
                        return read_tbx_impl<256, 8, T, N>(self, args);
                    case 12:
                        return read_tbx_impl<256, 12, T, N>(self, args);
                    case 16:
                        return read_tbx_impl<256, 16, T, N>(self, args);
                    default:
                        PyErr_Format(SyncError, "Mark 5C sync word differs from expected");
                        return NULL;
                }
            default:
                PyErr_Format(SyncError, "Mark 5C sync word differs from expected");
                return NULL;
        }
    }
    
    // Create the output data array
    npy_intp dims[3];
    // 4+4-bit Data -> 8192 samples in the data section as 16 channels, 256 stands, and 2 pols.
    dims[0] = (npy_intp) NCHAN;
    dims[1] = (npy_intp) NSTAND;
    dims[2] = (npy_intp) 2;
    if(N == NPY_INT8) dims[2] *= 2;
    data = (PyArrayObject*) PyArray_ZEROS(3, dims, N, 0);
    if(data == NULL) {
        PyErr_Format(PyExc_MemoryError, "Cannot create output array");
        Py_XDECREF(data);
        return NULL;
    }
    
    Py_BEGIN_ALLOW_THREADS
    
    // Swap the bits around
    cFrame.header.frame_count_word = __bswap_32(cFrame.header.frame_count_word);
    cFrame.header.second_count = __bswap_32(cFrame.header.second_count);
    cFrame.header.first_chan = __bswap_16(cFrame.header.first_chan);
    cFrame.payload.timetag = __bswap_64(cFrame.payload.timetag);
    
    // Fill the data array
    const int8_t *fp;
    T *a;
    a = (T *) PyArray_DATA(data);
    for(i=0; i<NCHAN*NSTAND*2*1; i+=4) {
        fp = tbxLUT[ cFrame.payload.bytes[i] ];
        *(a + 2*i + 0) = fp[0];
        *(a + 2*i + 1) = fp[1];
        fp = tbxLUT[ cFrame.payload.bytes[i+1] ];
        *(a + 2*i + 2) = fp[0];
        *(a + 2*i + 3) = fp[1];
        fp = tbxLUT[ cFrame.payload.bytes[i+2] ];
        *(a + 2*i + 4) = fp[0];
        *(a + 2*i + 5) = fp[1];
        fp = tbxLUT[ cFrame.payload.bytes[i+3] ];
        *(a + 2*i + 6) = fp[0];
        *(a + 2*i + 7) = fp[1];
    }
    
    Py_END_ALLOW_THREADS
    
    // Validate
    if( !validSync5C(cFrame.header.syncWord) ) {
        buffer = PyObject_CallMethod(ph, "seek", "ii", -frameSize, 1);
        PyErr_Format(SyncError, "Mark 5C sync word differs from expected");
        Py_XDECREF(data);
        return NULL;
    }
    
    // Save the data to the frame object
    // 1.  Header
    fHeader = PyObject_GetAttrString(frame, "header");
    
    temp = Py_BuildValue("B", cFrame.header.adp_id);
    PyObject_SetAttrString(fHeader, "adp_id", temp);
    Py_XDECREF(temp);
    
    temp = PyLong_FromUnsignedLong(cFrame.header.frame_count);
    PyObject_SetAttrString(fHeader, "frame_count", temp);
    Py_XDECREF(temp);
    
    temp = PyLong_FromUnsignedLong(cFrame.header.second_count);
    PyObject_SetAttrString(fHeader, "second_count", temp);
    Py_XDECREF(temp);
    
    temp = Py_BuildValue("h", cFrame.header.first_chan);
    PyObject_SetAttrString(fHeader, "first_chan", temp);
    Py_XDECREF(temp);
    
    temp = Py_BuildValue("h", nstand);
    PyObject_SetAttrString(fHeader, "nstand", temp);
    Py_XDECREF(temp);
    
    temp = Py_BuildValue("h", nchan);
    PyObject_SetAttrString(fHeader, "nchan", temp);
    Py_XDECREF(temp);
    
    // 2. Data
    fPayload = PyObject_GetAttrString(frame, "payload");
    
    temp = PyLong_FromLongLong(cFrame.payload.timetag);
    PyObject_SetAttrString(fPayload, "timetag", temp);
    Py_XDECREF(temp);
    
    PyObject_SetAttrString(fPayload, "_data", PyArray_Return(data));
    
    // 3. Frame
    PyObject_SetAttrString(frame, "header", fHeader);
    PyObject_SetAttrString(frame, "payload", fPayload);
    output = Py_BuildValue("O", frame);
    
    Py_XDECREF(fHeader);
    Py_XDECREF(fPayload);
    Py_XDECREF(data);
    
    return output;
}


template<typename T, NPY_TYPES N>
PyObject *read_tbx(PyObject *self, PyObject *args) {
    // Try fast path first with cached nstand
    switch(cached_nstand) {
        case 64:
            switch(cached_nchan) {
                case 4:
                    return read_tbx_impl<64, 4, T, N>(self, args);
                case 8:
                    return read_tbx_impl<64, 8, T, N>(self, args);
                case 12:
                    return read_tbx_impl<64, 12, T, N>(self, args);
                case 16:
                default:
                    return read_tbx_impl<64, 16, T, N>(self, args);
            }
        case 128:
            switch(cached_nchan) {
                case 4:
                    return read_tbx_impl<128, 4, T, N>(self, args);
                case 8:
                    return read_tbx_impl<128, 8, T, N>(self, args);
                case 12:
                    return read_tbx_impl<128, 12, T, N>(self, args);
                case 16:
                default:
                    return read_tbx_impl<128, 16, T, N>(self, args);
            }
        case 256: 
        default:
            switch(cached_nchan) {
                case 4:
                    return read_tbx_impl<256, 4, T, N>(self, args);
                case 8:
                    return read_tbx_impl<256, 8, T, N>(self, args);
                case 12:
                    return read_tbx_impl<256, 12, T, N>(self, args);
                case 16:
                default:
                    return read_tbx_impl<256, 16, T, N>(self, args);
            }
    }
}


PyObject *read_tbx_cf32(PyObject *self, PyObject *args) {
    return read_tbx<float,NPY_COMPLEX64>(self, args);
}

char read_tbx_cf32_doc[] = PyDoc_STR(\
"Function to read in a single TBW frame (header+data) and store the contents\n\
as a Frame object.\n\
\n\
.. versionadded:: 1.2.0\n\
");


PyObject *read_tbx_ci8(PyObject *self, PyObject *args) {
    return read_tbx<int8_t,NPY_INT8>(self, args);
}

char read_tbx_ci8_doc[] = PyDoc_STR(\
"Function to read in a single TBW frame (header+data) and store the contents\n\
as a Frame object.\n\
\n\
.. note::\n\
    This function differs from `read_tbx` in that it returns a\n\
    `lsl.tbx.FramePayload` containing a 4-D numpy.int8 array (channels by\n\
    stands by polarizations by real/complex) rather than a 1-D numpy.complex64\n\
    array.\n\
\n\
.. versionadded:: 2.1.3\n\
");

File: lsl/reader/tbx.py
"""
Python module to reading in data from TBX files.  TBX data are a complex
frequency-domain product that contains blocks of up to 16 channels from all
antennas in the array.  Each channel has a bandwidth of f\ :sub:`C` (25 kHz)
and there may be up to 224 different blocks of channels within a single
recording.  The stand ordering is based on the input into the digital system
rather than the stand number in the array.

This module defines the following classes for storing the TBX data found in a
file:

Frame
  object that contains all data associated with a particular TBX frame.  The 
  primary consituents of each frame are:
    * FrameHeader - the TBX frame header object and
    * FramePayload   - the TBX frame data object.  
Combined, these two objects contain all of the information found in the 
original TBX frame.

The functions defined in this module fall into two class:
  1. convert a frame in a file to a Frame object and
  2. describe the format of the data in the file.

For reading in data, use the read_frame function.  It takes a python file-
handle as an input and returns a fully-filled Frame object.

.. versionadded:: 3.1.0
"""

import numpy as np

from lsl.common import adp as adp_common
from lsl.common import ndp as ndp_common
from lsl.reader.base import *
from lsl.reader._gofast import read_tbf, read_tbf_ci8
from lsl.reader._gofast import SyncError as gSyncError
from lsl.reader._gofast import EOFError as gEOFError
from lsl.reader.errors import SyncError, EOFError
from lsl.reader.utils import FilePositionSaver

from lsl.misc import telemetry
telemetry.track_module()


__version__ = '0.1'
__all__ = ['FrameHeader', 'FramePayload', 'Frame', 'read_frame', 'read_frame_ci8',
           'FRAME_CHANNEL_COUNT', 'get_frame_size', 'get_frames_per_obs',
           'get_first_frame_count', 'get_channel_count', 'get_first_channel']

#: Number of frequency channels in a TBF packet
FRAME_CHANNEL_COUNT = 12


class FrameHeader(FrameHeaderBase):
    """
    Class that stores the information found in the header of a TBF 
    frame.  All three fields listed in the DP ICD version H are stored as 
    well as the original binary header data.
    """
    
    _header_attrs = ['frame_count', 'second_count', 'nstand', 'nchan', 'first_chan']
    
    def __init__(self, frame_count=None, second_count=None, first_chan=None, nstand=None, nchan=None):
        self.frame_count = frame_count
        self.second_count = second_count
        self.first_chan = first_chan
        self.nstand = nstand
        self.nchan = nchan
        FrameHeaderBase.__init__(self)
        
    @property
    def channel_freqs(self):
        """
        Return a numpy.float32 array for the center frequencies, in Hz, of
        each channel in the data.
        """
        
        fC = ndp_common.fC
        return (np.arange(FRAME_CHANNEL_COUNT, dtype=np.float32)+self.first_chan) * fC


class FramePayload(FramePayloadBase):
    """
    Class that stores the information found in the data section of a TBF
    frame.  Both fields listed in the DP ICD version H are stored.
    """
    
    _payload_attrs = ['timetag']
    
    def __init__(self, timetag=None, fDomain=None):
        self.timetag = timetag
        FramePayloadBase.__init__(self, fDomain)
        
    @property
    def time(self):
        """
        Function to convert the time tag from samples since the UNIX epoch
        (UTC 1970-01-01 00:00:00) to seconds since the UNIX epoch as a 
        `lsl.reader.base.FrameTimestamp` instance.
        """
        
        return FrameTimestamp.from_dp_timetag(self.timetag)


class Frame(FrameBase):
    """
    Class that stores the information contained within a single TBF 
    frame.  It's properties are FrameHeader and FramePayload objects.
    """
    
    _header_class = FrameHeader
    _payload_class = FramePayload
    
    @property
    def nstand(self):
        """
        Convenience wrapper for the Frame.FrameHeader.nstand property.
        """
        
        return self.header.nstand
        
    @property
    def nchan(self):
        """
        Convenience wrapper for the Frame.FrameHeader.nchan property.
        """
        
        return self.header.nchan
        
    @property
    def channel_freqs(self):
        """
        Convenience wrapper for the Frame.FrameHeader.channel_freqs property.
        """
        
        return self.header.channel_freqs
        
    @property
    def time(self):
        """
        Convenience wrapper for the Frame.FramePayload.time property.
        """
        
        return self.payload.time


def read_frame(filehandle, verbose=False):
    """
    Function to read in a single TBF frame (header+data) and store the 
    contents as a Frame object.
    """
    
    # New Go Fast! (TM) method
    try:
        newFrame = read_tbx(filehandle, Frame())
    except gSyncError:
        mark = filehandle.tell()
        raise SyncError(location=mark)
    except gEOFError:
        raise EOFError
        
    return newFrame


def read_frame_ci8(filehandle, verbose=False):
    """
    Function to read in a single TBX frame (header+data) and store the 
    contents as a Frame object.
    
    .. note::
        This function differs from `read_frame` in that it returns a
        `lsl.reader.tbf.FramePayload` that contains a 4-D numpy.int8 array
        (channels by stands by polarizations by by real/complex) rather than a
        3-D numpy.complex64 array.
    
    .. versionadded:: 2.1.3
    """
    
    # New Go Fast! (TM) method
    try:
        newFrame = read_tbx_ci8(filehandle, Frame())
        newFrame.payload._data = newFrame.payload.data.view(CI8)
    except gSyncError:
        mark = filehandle.tell() - FRAME_SIZE
        raise SyncError(location=mark)
    except gEOFError:
        raise EOFError
        
    return newFrame


def get_frame_size(filehandle):
    """
    Find out what the frame size in a file is at the current file location.
    Returns the frame size in bytes.
    """
    
    nPos = -1
    with FilePositionSaver(filehandle):
        for i in range(2500):
            try:
                cPos = filehandle.tell()
                cFrame = read_frame(filehandle)
                if not cFrame.is_tbf:
                    continue
                nPos = filehandle.tell()
                break
            except EOFError:
                break
            except SyncError:
                filehandle.seek(1, 1)
                
    if nPos < 0:
        raise RuntimeError("Unable to determine a frame size")
        
    return nPos - cPos


def get_frames_per_obs(filehandle):
    """
    Find out how many frames are present per time stamp by examining the 
    first 2500 TBF records.  Return the number of frames per observation.
    """
    
    with FilePositionSaver(filehandle):
        # Build up the list-of-lists that store the index of the first frequency
        # channel in each frame.
        channels = []
        for i in range(2500):
            try:
                cFrame = read_frame(filehandle)
                if not cFrame.is_tbf:
                    continue
            except EOFError:
                break
            except SyncError:
                filehandle.seek(0, 1)
                continue
                
            chan = cFrame.header.first_chan
            if chan not in channels:
                channels.append( chan )
                
    # Return the number of channels
    return len(channels)


def get_first_frame_count(filehandle):
    """
    Find and return the lowest frame count encountered in a TBF file.
    """
    
    # Find out how many frames there are per observation
    nFrames = get_frames_per_obs(filehandle)
    
    with FilePositionSaver(filehandle):
        firstFrameCount = 2**64-1
        freqs = []
        while len(freqs) < nFrames:
            cFrame = read_frame(filehandle)
            freq = cFrame.header.first_chan
            
            if freq not in freqs:
                freqs.append(freq)
            if cFrame.header.frame_count < firstFrameCount:
                firstFrameCount = cFrame.header.frame_count
                
    # Return the lowest frame number found
    return firstFrameCount


def get_channel_count(filehandle):
    """
    Find out the total number of channels that are present by examining 
    the first 1000 TBF records.  Return the number of channels found.
    """
    
    # Find out how many frames there are per observation
    nFrames = get_frames_per_obs(filehandle)
    
    # Convert to channels
    nChannels = nFrames * FRAME_CHANNEL_COUNT
    
    # Return the number of channels
    return nChannels


def get_first_channel(filehandle, frequency=False, all_frames=False):
    """
    Find and return the lowest frequency channel in a TBF file.  If the 
    `frequency` keyword is True the returned value is in Hz.  If `all` is
    True then the lowest frequency in each unique TBF frame is returned as
    a list.
    """
    
    # Find out how many frames there are per observation
    nFrames = get_frames_per_obs(filehandle)
    
    with FilePositionSaver(filehandle):
        # Find the lowest frequency channel
        freqs = []
        while len(freqs) < nFrames:
            cFrame = read_frame(filehandle)
            if frequency:
                freq = cFrame.channel_freqs[0]
            else:
                freq = cFrame.header.first_chan
                
            if freq not in freqs:
                freqs.append(freq)
                
    freqs.sort()
    if all_frames:
        # Return all unique first frequency channels
        return freqs
    else:
        # Return the lowest frequency channel
        return freqs[0]
